### PR TITLE
Correctly terminate MPI

### LINKED ISSUE #356: Use MPI Abort() call on error
We should try to abort the MPI job by calling `comm.Abort()` when one of the nodes decides it needs to terminate.

### PR DESCRIPTION
**Pull request by andrrizzi**
Fix #356 and avoid ncfile data corruption.

### CODE REVIEW

**Comment by andrrizzi** [comment]
Ready to be merged!

I've tested this method on a minimal example on the cluster and it seems to work well.

**Comment by jchodera** [comment]
Testing now!

**Comment by jchodera** [comment]
I think we may need to cut a new release of `openmoltools`:
```
[chodera@gpu-2-16 p-xylene-implicit]$ mpirun -configfile configfile

Traceback (most recent call last):
  File "/cbio/jclab/home/chodera/miniconda/bin/yank", line 9, in <module>
    load_entry_point('yank==0.9.0', 'console_scripts', 'yank')()
  File "/cbio/jclab/home/chodera/miniconda/lib/python2.7/site-packages/setuptools-23.0.0-py2.7.egg/pkg_resources/__init__.py", line 542, in load_entry_point
  File "/cbio/jclab/home/chodera/miniconda/lib/python2.7/site-packages/setuptools-23.0.0-py2.7.egg/pkg_resources/__init__.py", line 2569, in load_entry_point
  File "/cbio/jclab/home/chodera/miniconda/lib/python2.7/site-packages/setuptools-23.0.0-py2.7.egg/pkg_resources/__init__.py", line 2229, in load
  File "/cbio/jclab/home/chodera/miniconda/lib/python2.7/site-packages/setuptools-23.0.0-py2.7.egg/pkg_resources/__init__.py", line 2235, in resolve
  File "/cbio/jclab/home/chodera/miniconda/lib/python2.7/site-packages/yank/__init__.py", line 19, in <module>
    from yank import Yank
  File "/cbio/jclab/home/chodera/miniconda/lib/python2.7/site-packages/yank/yank.py", line 33, in <module>
    from sampling import ModifiedHamiltonianExchange # TODO: Modify to 'from yank.sampling import ModifiedHamiltonianExchange'?
  File "/cbio/jclab/home/chodera/miniconda/lib/python2.7/site-packages/yank/sampling.py", line 28, in <module>
    from repex import ThermodynamicState
  File "/cbio/jclab/home/chodera/miniconda/lib/python2.7/site-packages/yank/repex.py", line 74, in <module>
    from utils import is_terminal_verbose, delayed_termination
  File "/cbio/jclab/home/chodera/miniconda/lib/python2.7/site-packages/yank/utils.py", line 21, in <module>
    from openmoltools.utils import wraps_py2, unwrap_py2  # Shortcuts for other modules
ImportError: cannot import name wraps_py2
```

**Comment by jchodera** [comment]
Looks like `openmoltools-dev` hasn't been recently updated either:
https://anaconda.org/omnia/openmoltools/files

**Comment by andrrizzi** [comment]
Yes! That's a function that I've recently pushed. I've updated the `yank-dev` recipe to use `openmoltools-dev`.

**Comment by jchodera** [comment]
I installed [`openmoltools-dev`](https://anaconda.org/omnia/openmoltools-dev/files), but it apparently doesn't contain the `wraps_py2` code:
```
[chodera@gpu-2-16 ~]$ conda install --yes openmoltools-dev
...
[chodera@gpu-2-16 ~]$ python
from Python 2.7.11 |Continuum Analytics, Inc.| (default, Dec  6 2015, 18:08:32) 
[GCC 4.4.7 20120313 (Red Hat 4.4.7-1)] on linux2
Type "help", "copyright", "credits" or "license" for more information.
Anaconda is brought to you by Continuum Analytics.
Please check out: http://continuum.io/thanks and https://anaconda.org
>>> from openmoltools.utils import wraps_py2
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: cannot import name wraps_py2
```

**Comment by andrrizzi** [comment]
Ah! Weird. Let me look into that. Meanwhile, can you test by installing `openmoltools` from `setup.py`?

**Comment by andrrizzi** [comment]
You have probably already done this. But I would installed `yank-dev` from scratch on a separate environment. I had to update all packages back to the `dev` versions because I was having weird conflicts recently. Maybe that messed up your conda setup.

**Comment by jchodera** [comment]
I can confirm that this works with manually-installed `openmoltools`! So this PR should be good to merge.

> You have probably already done this. But I would installed yank-dev from scratch on a separate environment. I had to update all packages back to the dev versions because I was having weird conflicts recently. Maybe that messed up your conda setup.

Let me try again. I thought I was ultra-paranoid about removing everything, though!

**Comment by andrrizzi** [comment]
Awesome!

Travis tests should not pass if `openmoltools-dev` doesn't contain `wraps_py2` because I import it in `utils`. If tests pass in spite of that, it means there is something I don't understand I have to check.

**Comment by jchodera** [comment]
I'm still having no luck after manually cleaning out `openmoltools` and installing on `linux` via `conda install --yes openmoltools-dev`. Which PR was this pushed in?

**Comment by jchodera** [comment]
Hm, I see it was pushed 13 days ago, and there was an updated `openmoltools-dev` package built just a few hours ago.  Not sure what is going on.

**Comment by jchodera** [comment]
OK, this works for @rafwiewiora too. I'll see what is going on with my install...

Merge when ready!

**Comment by andrrizzi** [comment]
Great! Merging.
